### PR TITLE
[ECO-4943] feat: basic sending and receiving messages in the Chat

### PR DESCRIPTION
**Pull request by ttypic**
Resolves https://github.com/ably-labs/ably-chat-kotlin/issues/2

In this PR, we introduce the basic implementation for sending and receiving messages. Corner cases and potential race conditions are not the focus at this stage and will be addressed later, once Room lifecycle management is implemented and a unified test suite is set up.

## Release Notes

- **New Features**
  - Enhanced message retrieval with optional parameters for more granular control.
  - Introduced new utility functions for JSON handling, improving robustness.
  - Simplified chat functionality by transitioning to `ChatClient` for message handling.

- **Improvements**
  - Default parameters added to several functions for easier usage.
  - Improved encapsulation and resource management in room handling.
  - Enhanced test suite for messaging functionality, ensuring reliability.

- **Refactor**
  - Streamlined test setup code for better readability and maintainability.
  - Refactored message subscription logic for improved user experience.
  - Introduced a new mechanism for deferred value handling in utility functions.
  - Updated internal structure of various classes to enhance maintainability.

### CODE REVIEW

File: chat-android/src/main/java/com/ably/chat/Messages.kt [hunk 1]
```diff
@@ -170,25 +180,96 @@ data class SendMessageParams(
...
 
-    override val channel: Channel
-        get() = realtimeClient.channels.get(messagesChannelName, ChatChannelOptions())
+    override val channel: Channel = realtimeChannels.get(messagesChannelName, ChatChannelOptions())
+
+    init {
+        channelStateListener = ChannelStateListener {
+            if (!it.resumed) updateChannelSerialsAfterDiscontinuity()
+        }
+        channel.on(ChannelState.attached, channelStateListener)
```
Comment (AndyTWF) [inline chat-android/src/main/java/com/ably/chat/Messages.kt:230]: We also need to handle `ChannelEvent.update` here - which is RTL2g in the core feature spec

File: chat-android/src/main/java/com/ably/chat/Messages.kt [hunk 2]
```diff
@@ -170,25 +180,96 @@ data class SendMessageParams(
...
 
     override fun subscribe(listener: Messages.Listener): MessagesSubscription {
-        TODO("Not yet implemented")
+        val deferredChannelSerial = DeferredValue<String>()
+        addListener(listener, deferredChannelSerial)
+        val messageListener = PubSubMessageListener {
+            val pubSubMessage = it ?: throw AblyException.fromErrorInfo(
+                ErrorInfo("Got empty pubsub channel message", HttpStatusCodes.BadRequest, ErrorCodes.BadRequest),
+            )
+            val data = parsePubSubMessageData(pubSubMessage.data)
```
Comment (AndyTWF) [inline chat-android/src/main/java/com/ably/chat/Messages.kt:240]: Possible optimisation we can do here - rather than parsing data for every subscription, can have an "internal" subscriber that's put on the channel which then translates once and emits to all "actual" subscribers.

In JS we did this by registering the "internal" subscriber with the realtime channel, then registering user listeners with an `EventEmitter` instance on this class.

File: chat-android/src/main/java/com/ably/chat/Messages.kt [hunk 3]
```diff
@@ -170,25 +180,96 @@ data class SendMessageParams(
...
+                clientId = pubSubMessage.clientId,
+                timeserial = pubSubMessage.extras.asJsonObject().requireString("timeserial"),
+                text = data.text,
+                metadata = data.metadata,
+                headers = pubSubMessage.extras.asJsonObject().get("headers")?.toMap() ?: mapOf(),
+            )
+            emitMessage(chatMessage)
+        }
+
+        channel.subscribe(MessageEventType.Created.eventName, messageListener)
```
Comment (AndyTWF) [inline chat-android/src/main/java/com/ably/chat/Messages.kt:253]: This would cause an implicit channel attach which will break room lifecycle - subscribing should cause no side-effects.

I assume there's an ably-java ticket that needs doing first? :)

File: chat-android/src/main/java/com/ably/chat/Messages.kt (hunk 2 above)
Comment (AndyTWF) [reply chat-android/src/main/java/com/ably/chat/Messages.kt:240]: Looking at code further down, won't this cause duplicate messages on multiple subscriptions?

As multiple instances of `messageListener` will call `emitMessage` which iterates all listeners?

File: chat-android/src/main/java/com/ably/chat/Messages.kt [hunk 4]
```diff
@@ -170,25 +180,96 @@ data class SendMessageParams(
...
+        val deferredChannelSerial = DeferredValue<String>()
+        addListener(listener, deferredChannelSerial)
+        val messageListener = PubSubMessageListener {
+            val pubSubMessage = it ?: throw AblyException.fromErrorInfo(
+                ErrorInfo("Got empty pubsub channel message", HttpStatusCodes.BadRequest, ErrorCodes.BadRequest),
+            )
+            val data = parsePubSubMessageData(pubSubMessage.data)
+            val chatMessage = Message(
+                roomId = roomId,
+                createdAt = pubSubMessage.timestamp,
```
Comment (AndyTWF) [inline chat-android/src/main/java/com/ably/chat/Messages.kt:243]: In JS we've tended towards native date types for `createdAt` - not sure what's better for Kotlin?

File: .gitignore [hunk 5]
```diff
@@ -96,6 +96,7 @@ publish.properties
 /.idea/jarRepositories.xml
 /.idea/misc.xml
 /.idea/shelf
+/.idea/uiDesigner.xml
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply .gitignore:99]: Yes, IDE specific settings should be ignored `.idea/**`

[the remark this answers is not in the document]
Comment (ttypic) [reply .gitignore:99]: No, we shouldn't ignore all files in `.idea` folder. It contains dictionary and codestyles settings that are useful if you use jetbrains IDE (Android Studio or IDEA). It's a good practice to add this settings for kotlin projects. But I obviously copied really bad gitignore pattern from some kotlin library template, didn't put much thought to it. I'll remove changes to gitignore in this PR and add separate PR that introduce updates to gitignore

File: chat-android/src/main/java/com/ably/chat/Messages.kt (hunk 1 above)
Comment (ttypic) [reply chat-android/src/main/java/com/ably/chat/Messages.kt:230]: Thanks, nice catch!

File: chat-android/src/main/java/com/ably/chat/Messages.kt (hunk 2 above)
Comment (ttypic) [reply chat-android/src/main/java/com/ably/chat/Messages.kt:240]: Yes, you right, I'll add tests :) I was thinking to use internal eventbus, but decided to do this later, because we need to decide on implementation details. I'll create an issue for this

File: chat-android/src/main/java/com/ably/chat/Messages.kt (hunk 3 above)
Comment (ttypic) [reply chat-android/src/main/java/com/ably/chat/Messages.kt:253]: Yes, of course, https://github.com/ably/ably-java/issues/1027, I think until it release we can leave it as is. I'll add separate issue for this, too

File: chat-android/src/main/java/com/ably/chat/Messages.kt (hunk 4 above)
Comment (ttypic) [reply chat-android/src/main/java/com/ably/chat/Messages.kt:243]: I don't have strong opinion right now. Kotlin standard library doesn't have date type out of the box, and I still believe that one day we can provide  Kotlin multiplatform support for all our products :) I'll think about this, look how easy to use timestamps in example app. Also want to research how libraries with kotlin multiplatform support like apollo-graphql deals with this. For now I suggest keeping date as `Long`

Comment (sacOO7) [reply chat-android/src/main/java/com/ably/chat/Messages.kt:243]: wdyt about https://github.com/Kotlin/kotlinx-datetime. It's a multiplatform datetime library

Comment (ttypic) [reply chat-android/src/main/java/com/ably/chat/Messages.kt:243]: If I’m not mistaken, it’s not stable yet, so I don’t think we should depend on it right now. We can decide this later—it’s not an architectural design decision that will cause issues for us later, and it will be easy to refactor whenever we want.

File: chat-android/src/main/java/com/ably/chat/Messages.kt (hunk 1 above)
Comment (ttypic) [reply chat-android/src/main/java/com/ably/chat/Messages.kt]: fixed

File: chat-android/src/main/java/com/ably/chat/Messages.kt (hunk 2 above)
Comment (ttypic) [reply chat-android/src/main/java/com/ably/chat/Messages.kt:240]: https://github.com/ably-labs/ably-chat-kotlin/issues/18

File: chat-android/src/main/java/com/ably/chat/Messages.kt (hunk 3 above)
Comment (ttypic) [reply chat-android/src/main/java/com/ably/chat/Messages.kt:253]: https://github.com/ably-labs/ably-chat-kotlin/issues/17

**Review by AndyTWF — APPROVED** [review]

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (ttypic) [reply .gitignore]: https://github.com/ably-labs/ably-chat-kotlin/pull/19/files

File: example/src/main/java/com/ably/chat/example/MainActivity.kt [hunk 6]
```diff
@@ -46,20 +46,23 @@ val randomClientId = UUID.randomUUID().toString()
 class MainActivity : ComponentActivity() {
```
Comment (sacOO7) [inline example/src/main/java/com/ably/chat/example/MainActivity.kt:46]: I do not support pure kotlin code for UI, instead it should be put into layout.xml files.
This is more like writing website using pure js without use of html + css. 
One of the biggest drawback is lack of local preview when adjusting UI components. Right now, we need to rerun app to look for breaking UI components. Also, you can't preview layout on different devices.
I am assuming this is a temporary code and will need complete rewrite in the near future ( Since code is not clean and is a mix of UI + business code )

**Review by sacOO7 — APPROVED** [review]
Overall looks good. More tests can be added in the future. 
Also, need to update example app with clean code as per comment https://github.com/ably-labs/ably-chat-kotlin/pull/16#discussion_r1763040012

File: example/src/main/java/com/ably/chat/example/MainActivity.kt (hunk 6 above)
Comment (ttypic) [reply example/src/main/java/com/ably/chat/example/MainActivity.kt:47]: No more xml in Android, it's old news :) with Jetpack Compose everything is in Kotlin, there are special preview component for UI. I added them.

**Review by AndyTWF — APPROVED** [review]

File: example/src/main/java/com/ably/chat/example/MainActivity.kt (hunk 6 above)
Comment (sacOO7) [reply example/src/main/java/com/ably/chat/example/MainActivity.kt:47]: Wow, I didn't know so many things changed. I loved working with xml. Lots of threads talking about the same
https://www.reddit.com/r/androiddev/comments/1bbbsop/why_are_people_against_xml_now/

Comment (sacOO7) [reply example/src/main/java/com/ably/chat/example/MainActivity.kt:47]: Okay, I will check new way to develop UI layouts, thanks!

Comment (sacOO7) [reply example/src/main/java/com/ably/chat/example/MainActivity.kt:47]: https://medium.com/@MeenoTeK/xml-views-or-jetpack-compose-which-is-the-best-option-for-your-next-project-ccf38573a82

File: example/src/main/java/com/ably/chat/example/MainActivity.kt [hunk 7]
```diff
@@ -69,29 +73,42 @@
...
+        coroutineScope.launch {
+            messages = subscription.getPreviousMessages().items.reversed()
+            if (messages.isNotEmpty()) listState.animateScrollToItem(messages.size - 1)
+        }
+
+        onDispose {
+            subscription.unsubscribe()
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (ttypic) [reply example/src/main/java/com/ably/chat/example/MainActivity.kt:103]: There is no method `awaitDispose` in the `CoroutineScope`. `coroutineScope` in the code is produced by `rememberCoroutineScope` function, so it can't outlive the composable's lifecycle

[the remark this answers is not in the document]
Comment (sacOO7) [reply example/src/main/java/com/ably/chat/example/MainActivity.kt:103]: Can you make the face you made again 🙉

[the remark this answers is not in the document]
Comment (sacOO7) [reply example/src/main/java/com/ably/chat/example/MainActivity.kt:103]: Nah it was for you @coderabbitai

[the remark this answers is not in the document]
Comment (sacOO7) [reply example/src/main/java/com/ably/chat/example/MainActivity.kt:103]: Can you describe what it exactly represents @coderabbitai

[the remark this answers is not in the document]
Comment (sacOO7) [reply example/src/main/java/com/ably/chat/example/MainActivity.kt:103]: Can you add to the learnings to make happy face when you feel empathetic next time
